### PR TITLE
http-status-codes v2

### PR DESCRIPTION
**Pull request by prettymuchbryce**
This library could use some major improvements. Even though this project is extremely narrow in scope, the JavaScript ecosystem has evolved drastically since it was first written in 2013. There are few major glaring problems that are affecting the development speed of this project.

## Testing

As mentioned in #46, There are no tests to speak of in this library currently. The initial scope of this project was so small that I did not think it was necessary, but a consequence of the lack of testing is that it makes it difficult for me to review and approve changes with confidence. Even simple changes like adding a new code give me hesitation because breaking any users is completely unacceptable.

One of the improvements along these lines in v2 is using `ts-morph` to generate all of the source related to the status codes themselves. This should make adding new status codes much easier and safer. I want to make it easy for anyone to suggest new codes without risking breaking any source code.

## TypeScript Support

It's clear that a lot of users of this library are using TypeScript now and they want to make the most of its features (like Enums in #43). This library doesn't play very nice with such features currently. Additionally, we have been effectively maintaining two separate implementations of this library rather than just generating the JavaScript from TypeScript.

## Backwards Compatibility

While I want to maintain backwards compatibility with v1 for now, I would still prefer to make this change opt-in by bumping the major version to v2. I've added some tests here that also test for compatibility with v1.

My hope is that some of the improvements here will help solve some of the pain points we've been feeling. A huge thanks to @seanmhanson who inspired a lot of these changes with their great suggestions.

### CODE REVIEW

**Comment by seanmhanson** [comment]
Wow. This is great progress and stuff I was thinking of suggesting re: testing etc (testing is one of my favorite things to focus on). If you'd like, I can do a quick pass-through to help out the review process, just in case anything jumps out. Also, if you don't mind a quick edit, I use they/them pronouns :)

**Comment by prettymuchbryce** [comment]
@seanmhanson Apologies for that. Fixed.

Any review would definitely be appreciated. Let me know if you have any thoughts or ideas!

File: README.md [hunk 1]
```diff
@@ -15,122 +15,136 @@ npm install http-status-codes --save
...
+```
+These changes are __optional__ now, but may be required in a future major version so please consider making them now.
 
-response
-	.status(getStatusCode('Server Error')
-	.send('Server Error')
+
+## Proposing a new status code
+
+If you'd like to porpose a new status code, feel free to update "codes.json" with the necessary
```
Comment (seanmhanson) [inline README.md:136]: [typo] `porpose` -> `propose`

File: package.json [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
   "name": "http-status-codes",
-  "version": "1.4.0",
+  "version": "2.0.0-beta.0",
   "description": "Constants enumerating the HTTP status codes. Based on the Java Apache HttpStatus API.",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "update-codes": "ts-node --project ./scripts/tsconfig.json ./scripts/update-codes",
+    "build-ts": "tsc --project tsconfig-build.json",
+    "build": "npm run build-ts",
+    "test": "jest ./src"
```
Comment (seanmhanson) [inline package.json:9]: [no action] 😍   it's so good to see Jest!

File: package.json [hunk 3]
```diff
@@ -21,5 +24,19 @@
...
+    "eslint": "^7.7.0",
+    "eslint-config-airbnb-typescript": "^9.0.0",
+    "eslint-plugin-import": "^2.22.0",
+    "fs-extra": "^9.0.1",
+    "jest": "^26.4.0",
+    "markdown-table": "^2.0.0",
+    "ts-jest": "^26.2.0",
+    "ts-morph": "^7.3.0",
+    "typescript": "^3.9.7",
+    "@types/fs-extra": "^9.0.1"
```
Comment (seanmhanson) [inline package.json:40]: [nitpick] `@type` dependencies should be grouped together

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "module": "es6",
+    "target": "es5",
```
Comment (seanmhanson) [inline tsconfig.json:4]: [no action] 👍🏻   preserving ES5 compatibility. It's so weird to me to think that ES6 is mostly just assumed now 👴🏻

File: README.md [hunk 5]
```diff
@@ -15,122 +15,136 @@ npm install http-status-codes --save
...
+information and open a pull request. There is no need to modify any source code as it is all
```
Comment (seanmhanson) [inline README.md:137]: [nitpick] I'd throw in that the `README.md` is also updated automatically (I initially commented this, then read the script and realized you took care of this already, which is awesome).

Maybe something like this, though my tone is less formal:
> No need to modify source code or even this README–the `update-codes` script will handle this all for you.

File: scripts/update-codes.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+const run = async () => {
+  console.log('Updating src/codes.ts');
+
+  const project = new Project({
+    tsConfigFilePath: 'tsconfig.json',
+  });
+
+  // TODO @prettymuchbryce clean up these horribly unreadable one-liners
```
Comment (seanmhanson) [inline scripts/update-codes.ts:37]: [suggestion] Here's a start, at the least, that you can tweak to your preferences. You could take it a step further and pull out `OptionalKind<EnumMemberStucture>[]` into a type, and even `Record<string, string>` and `Record<string, number`, though neither bother me. I find logic in template literals can become a bit hard to read quickly, so I pulled out any logic, but you don't need to.

```
const reasonPhraseMembers: OptionalKind<EnumMemberStructure>[] = Codes
    .map(({ phrase, constant, comment, isDeprecated }: jsonCode) => {
        const { doc, description } = comment;
        const deprecatedString = isDeprecated ? '@deprecated\n': '';
        return {
            name: constant,
            value: phrase,
            docs: [`${deprecatedString}${doc}\n\n${description}`]
        };
    });

const statusCodeMembers: OptionalKind<EnumMemberStructure>[] = Codes
    .map(({ code, constant, comment, isDeprecated }: jsonCode) => {
        const { doc, description } = comment;
        const deprecatedString = isDeprecated ? '@deprecated\n': '';
        return {
            name: constant,
            value: code,
            docs: [`${deprecatedString}${doc}\n\n${description}`]
        };
    });

const statusCodeToReasonPhrase = Codes
    .reduce((acc: Object, { code, phrase }) => { 
        (acc as Record<string, string>)[`"${code.toString()}"`] = `"${phrase}"`;
        return acc;
    }, {});

const reasonPhraseToStatusCode = Codes
    .reduce((acc: Object, { code, phrase }) => {
        (acc as Record<string, number>)[`"${phrase}"`] = code;
        return acc;
    }, {});
```

File: README.md [hunk 7]
```diff
@@ -15,122 +15,136 @@ npm install http-status-codes --save
...
+generated based on this file.
+
+In general, we try to include only codes that have an official RFC and have been approved, however
+exceptions will be made if the code is already in widespread use in the wild.
+
+## Steps to build and publish
+
+```shell
+npm run update-codes
+npm run build
```
Comment (seanmhanson) [inline README.md:147]: [suggestion] I'm not sure where to put this comment, but I maybe you can add `npm run test`, or you can integrate running the tests to `update-codes`.

**Pull request merged**